ndcube/ndcube.py

```diff
@@ -4,74 +4,195 @@
-    def dimensions(self):
-        pass
+        When a slice is applied to a NDCube object it must be applied to the following attributes:
+
+          * data
+          * wcs
+          * uncertainty
+          * mask
+          * extra_coords
+
+        the following attributes must be copied unaltered to the returned NDCube:
+
+          * unit
+          * meta
+
+        """
 
     @abc.abstractmethod
-    def crop_by_coords(self, lower_corner, interval_widths=None, upper_corner=None, units=None):
+    @property
+    def extra_coords(self) -> "ExtraCoordsABC":
+        """
+        The extra coordinates associated with this cube.
+        """
+
+    @abc.abstractmethod
+    @property
+    def wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube.
+
+        If a `BaseLowLevelWCS` is provided to the constructor it should be
+        "upgraded" to a high level object.
+        """
+
+    @abc.abstractmethod
+    @property
+    def array_axis_physical_types(self) -> Iterable[Tuple[str, ...]]:
+        """
+        Returns the WCS physical types associated with each array axis.
+
+        Returns an iterable of tuples where each tuple corresponds to an array axis and
+        holds strings denoting the WCS physical types associated with that array axis.
+        Since multiple physical types can be associated with one array axis, tuples can
+        be of different lengths. Likewise, as a single physical type can correspond to
+        multiple array axes, the same physical type string can appear in multiple tuples.
+        """
+
+    @abc.abstractmethod
+    def axis_world_coords(self,
+                          *axes: Union[Union[int, str], Iterable[Union[int, str]]],
+                          edges: bool = False) -> Iterable[object]:
+        """
+        Returns world coordinate values of all pixels for all axes.
+
+        Parameters
+        ----------
+        axes
+            Axis number in numpy ordering or unique substring of
+            `.NDCube.wcs.world_axis_physical_types` or
+            `.NDCube.wcs.world_axis_names` of axes for which real world
+            coordinates are desired. ``axes=None`` implies all axes will be
+            returned.
+
+        edges
+            The edges argument helps in returning ``pixel_edges``
+            instead of ``pixel_values``. Default value is False,
+            which returns ``pixel_values``. True return ``pixel_edges``
+
+        Returns
+        -------
+        axes_coords
+            High level object, or iterable thereof, giving the real world
+            coords for the axes requested by user.
+            For example, SkyCoords. The types returned are determined by the WCS object.
+
+        Example
+        -------
+        >>> NDCube.all_world_coords(('lat', 'lon')) # doctest: +SKIP
+        >>> NDCube.all_world_coords(2) # doctest: +SKIP
+
+        """
+
+    @abc.abstractmethod
+    def axis_world_coords_values(self,
```

Comment:
Am I missing something but what is the major difference between `axis_world_coords` and `axis_world_coords_values` - its not super clear from the docstrings

Reply:
`axis_world_coords` returns high level objects, e.g. astropy `Time`, `SkyCoord`, `SpectralCoord`, etc. which may or may not combine world axes into single objects, e.g. lat, lon in a `SkyCoord`.  `axis_world_coords_values` returns a `Quantity` for each world axis.  So lat and lon are separate.  `axis_world_coords`'s object can have extra contextual information like what frame you're in.  While `_values` returns Quantities that may need extra information to interpret correctly, but can be easier to handle.